src/web/sound.rs

```diff
@@ -1,15 +1,143 @@
+use wasm_bindgen::{JsValue, JsCast};
+use js_sys::{Error, Array, ArrayBuffer, Map};
+use web_sys::{AudioContext, AudioBuffer, AudioNode, AudioBufferSourceNode, ConstantSourceNode, Response};
+use wasm_bindgen_futures::JsFuture;
 use futures::channel::mpsc::UnboundedReceiver;
+use futures::StreamExt;
+use webutil::event::EventTargetExt;
 
 use crate::sound::SoundCommand;
 
-// TODO
+pub(crate) type InternalSound = AudioBuffer;
 
-pub type InternalSound = ();
+thread_local! {
+    static AUDIO_CONTEXT: AudioContext = AudioContext::new().unwrap();
+}
 
 pub(crate) async fn load(source: &str) -> Result<InternalSound, String> {
-    Ok(())
+    async fn load_(source: &str) -> Result<InternalSound, JsValue> {
+        let response: Response = JsFuture::from(web_sys::window().unwrap().fetch_with_str(source))
+            .await?
+            .dyn_into()
+            .unwrap();
+        let buffer: ArrayBuffer = JsFuture::from(response.array_buffer()?)
+            .await?
+            .dyn_into()
+            .unwrap();
+        let buffer = JsFuture::from(AUDIO_CONTEXT.with(|ctx| ctx.decode_audio_data(&buffer))?)
+            .await?
+            .dyn_into()
+            .unwrap();
+        Ok(buffer)
+    }
+    load_(source).await.map_err(|err: JsValue| err.dyn_into::<Error>().unwrap().to_string().into())
+}
+
+enum ServiceState {
+    Playing(Map),
+    Paused(Vec<(AudioBuffer, f64)>)
+}
+
+struct SoundService {
+    gain_source: ConstantSourceNode
+}
+
+impl SoundService {
+    fn new() -> Self {
+        let gain_source = AUDIO_CONTEXT.with(|ctx| ctx.create_constant_source().unwrap());
+        gain_source.start().unwrap();
+
+        Self {
+            gain_source
+        }
+    }
+
+    fn set_volume(&self, volume: f32) {
+        self.gain_source.offset().set_value(volume);
+    }
+
+    fn play_sound(&self, ctx: &AudioContext, source_nodes: &Map, sound: &AudioBuffer, offset: f64) {
+        let gain = ctx.create_gain().unwrap();
+        gain.connect_with_audio_node(&ctx.destination()).unwrap();
+        self.gain_source.connect_with_audio_param(&gain.gain()).unwrap();
+
+        let source: AudioNode = ctx
+            .create_buffer_source()
+            .unwrap()
+            .dyn_into()
+            .unwrap();
+        source.connect_with_audio_node(&gain).unwrap();
+        let source: AudioBufferSourceNode = source
+            .dyn_into()
+            .unwrap();
+        source.set_buffer(Some(sound));
+        
+        wasm_bindgen_futures::spawn_local({
+            let source = source.clone();
+            let source_nodes = source_nodes.clone();
+            let event = source
+                .dyn_ref::<web_sys::EventTarget>()
+                .unwrap()
+                .once::<webutil::event::Ended>();
+
+            source_nodes.set(&source, &(ctx.current_time() - offset).into());
+            async move {
+                event.await;
+                source_nodes.delete(&source);
```

Comment:
To avoid this bit of data sharing and make the Rust compiler happy, you can use a channel here. In the sound service, you would have two streams, one of sound commands and one from here, and use `futures::select!` to query both at the same time.